libtenzir/builtins/formats/syslog.cpp

```diff
@@ -0,0 +1,117 @@
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/format/reader_factory.hpp>
+#include <tenzir/format/syslog.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/table_slice_builder.hpp>
+#include <tenzir/to_lines.hpp>
+
+namespace tenzir::plugins::syslog {
+
+namespace {
+
+auto impl(generator<std::optional<std::string_view>> lines,
+          operator_control_plane& ctrl) -> generator<table_slice> {
+  table_slice_builder syslog_builder{format::syslog::make_syslog_type()};
+  table_slice_builder legacy_syslog_builder{
+    format::syslog::make_legacy_syslog_type()};
+  table_slice_builder unknown_builder{format::syslog::make_unknown_type()};
+  table_slice_builder* builder = nullptr;
+  for (auto&& line : lines) {
+    // TODO: Flush builder if maximum batch size or timeout is reached?
+    if (not line) {
+      co_yield {};
+      continue;
+    }
+    if (line->empty()) {
+      co_yield {};
+      continue;
+    }
+    const auto* f = line->begin();
+    const auto* const l = line->end();
+    format::syslog::message msg{};
+    format::syslog::legacy_message legacy_msg{};
+    auto* const prev_builder = builder;
+    if (auto parser = format::syslog::message_parser{}; parser(f, l, msg)) {
+      builder = &syslog_builder;
+      if (not builder->add(msg.hdr.facility, msg.hdr.severity, msg.hdr.version,
+                           msg.hdr.ts, msg.hdr.hostname, msg.hdr.app_name,
+                           msg.hdr.process_id, msg.hdr.msg_id, msg.data,
+                           msg.msg)) {
+        diagnostic::error(
+          "syslog parser (RFC 5242) failed to produce table slice")
+          .hint(fmt::format("line: `{}`", *line))
+          .emit(ctrl.diagnostics());
+        co_return;
+      }
+    } else if (auto legacy_parser = format::syslog::legacy_message_parser{};
+               legacy_parser(f, l, legacy_msg)) {
+      builder = &legacy_syslog_builder;
+      if (not builder->add(legacy_msg.facility, legacy_msg.severity,
+                           legacy_msg.timestamp, legacy_msg.host,
+                           legacy_msg.tag, legacy_msg.content)) {
+        diagnostic::error(
+          "syslog parser (RFC 3164) failed to produce table slice")
+          .hint(fmt::format("line: `{}`", *line))
+          .emit(ctrl.diagnostics());
+        co_return;
+      }
+    } else {
+      builder = &unknown_builder;
+      if (not builder->add(*line)) {
+        diagnostic::error(
+          "syslog parser (unknown format) failed to produce table slice")
+          .hint(fmt::format("line: `{}`", *line))
+          .emit(ctrl.diagnostics());
+        co_return;
+      }
+    }
+    if (prev_builder != builder && prev_builder)
+      co_yield prev_builder->finish();
+  }
+  if (builder)
+    co_yield builder->finish();
```

Comment:
There is one more thing to consider here: If you get a long stream of let's say only RFC 5242 events, then you will only flush the builders when the incoming data ends. And if it is a continuous stream that does not end (e.g., TCP stream), then you will _never_ flush. We use 2 mechanisms to solve this problem: (1) A maximum batch size (`defaults::import::table_slice_size`), (2) and a maximum batch timeout (`defaults::import::batch_timeout`). You can take some inspiration from the other parsers.

Reply:
I did, in fact, take inspiration here from the CEF parser, where this is also not implemented, and where I copied the TODO comment from :smile: But sure, let's make this work properly without any TODOs

Reply:
I think the cleanest implementation of this is in the zeek-tsv parser if you're looking for inspiration

Reply:
NB: CEF over syslog is pretty common. Might be worth adding as integration test.

Reply:
I added a test case with a CEF payload. I couldn't really figure out how to actually parse the CEF as a part of the pipeline: `read syslog` gives me a record with the message in one of its fields, but I can't exactly pipe that to `read cef`. Might be something for the `parse` operator to do.

Reply:
> Might be something for the `parse` operator to do.

This. @jachris and I agreed on the UX for `parse` yesterday. The idea is that you'll be able to do:

```
load tcp://1.2.3.4
| read syslog
| parse message cef
```

Exact syntax TBD. The corollary is that `parse` takes a parser as argument, and that `kv`, `dissect`, and `grok` should also become parsers.